File: sdk/python/src/llm_observatory/instrumentors/anthropic.py
"""Anthropic auto-instrumentation."""

from __future__ import annotations

from typing import Any

from llm_observatory.pricing import calculate_cost
from llm_observatory.tracer import Span, SpanStatus, get_tracer


class AnthropicInstrumentor:
    """Instruments Anthropic API calls to capture traces."""

    def __init__(self) -> None:
        self._original_sync_create: Any | None = None
        self._original_async_create: Any | None = None

    def instrument(self) -> None:
        """Patch Anthropic to capture traces."""
        try:
            import anthropic

            # Patch sync messages
            self._original_sync_create = anthropic.resources.messages.Messages.create
            anthropic.resources.messages.Messages.create = self._patched_sync_create

            # Patch async messages
            self._original_async_create = (
                anthropic.resources.messages.AsyncMessages.create
            )
            anthropic.resources.messages.AsyncMessages.create = (
                self._patched_async_create
            )
        except ImportError:
            pass

    def _build_span(
        self, model: str, messages: list[dict[str, Any]]
    ) -> tuple[Span, Any]:
        """Build a span for an Anthropic call."""
        tracer = get_tracer()
        span = tracer.start_span(
            name="anthropic.messages.create",
            span_type="llm",
        )
        span.set_input({"model": model, "messages": messages})
        span.set_attribute("gen_ai.system", "anthropic")
        span.set_attribute("gen_ai.request.model", model)
        return span, tracer

    def _finish_span(self, span: Span, result: Any, model: str) -> None:
        """Extract data from result and finish span."""
        try:
            # Extract token usage
            if hasattr(result, "usage"):
                input_tokens = getattr(result.usage, "input_tokens", 0) or 0
                output_tokens = getattr(result.usage, "output_tokens", 0) or 0
                span.set_token_usage(input_tokens, output_tokens)

                # Calculate cost
                cost = calculate_cost(model, input_tokens, output_tokens)
                if cost is not None:
                    span.set_cost(cost)

            # Extract output
            if hasattr(result, "content") and result.content:
                content_blocks = []
                for block in result.content:
                    block_data = {}
                    if hasattr(block, "type"):
                        block_data["type"] = block.type
                    if hasattr(block, "text"):
                        block_data["text"] = block.text
                    content_blocks.append(block_data)
                span.set_output({"content": content_blocks})

            span.set_attribute("response_id", getattr(result, "id", None))
            span.set_status(SpanStatus.OK)

        except Exception as e:
            span.set_status(SpanStatus.ERROR)
            span.set_attribute("error.message", str(e))

        finally:
            span.end()

    def _patched_sync_create(self, *args: Any, **kwargs: Any) -> Any:
        """Patched sync create method."""
        model = kwargs.get("model", "unknown")
        messages = kwargs.get("messages", [])

        span, _ = self._build_span(model, messages)

        try:
            result = self._original_sync_create(*args, **kwargs)
            self._finish_span(span, result, model)
            return result
        except Exception as e:
            span.set_status(SpanStatus.ERROR)
            span.set_attribute("error.message", str(e))
            span.end()
            raise

    async def _patched_async_create(self, *args: Any, **kwargs: Any) -> Any:
        """Patched async create method."""
        model = kwargs.get("model", "unknown")
        messages = kwargs.get("messages", [])

        span, _ = self._build_span(model, messages)

        try:
            result = await self._original_async_create(*args, **kwargs)
            self._finish_span(span, result, model)
            return result
        except Exception as e:
            span.set_status(SpanStatus.ERROR)
            span.set_attribute("error.message", str(e))
            span.end()
            raise

    def uninstrument(self) -> None:
        """Remove instrumentation."""
        try:
            import anthropic

            if self._original_sync_create is not None:
                anthropic.resources.messages.Messages.create = (
                    self._original_sync_create
                )
            if self._original_async_create is not None:
                anthropic.resources.messages.AsyncMessages.create = (
                    self._original_async_create
                )
        except ImportError:
            pass

File: sdk/typescript/src/instrumentors/anthropic.ts
/**
 * Anthropic auto-instrumentation
 */

import { getTracer, type Span } from "../tracer";
import { calculateCost } from "../pricing";

let originalMessagesCreate: ((...args: unknown[]) => unknown) | null = null;
let originalAsyncMessagesCreate: ((...args: unknown[]) => unknown) | null = null;

function buildSpan(model: string, messages: unknown[]): Span {
  const tracer = getTracer();
  const span = tracer.startSpan("anthropic.messages.create", {
    spanType: "llm",
    attributes: {
      "gen_ai.system": "anthropic",
      "gen_ai.request.model": model,
      model,
      messages,
    },
  });
  span.input = { model, messages };
  return span;
}

function finishSpan(span: Span, result: unknown, model: string): void {
  if (result && typeof result === "object") {
    const obj = result as Record<string, unknown>;

    // Extract token usage
    const usage = obj.usage as Record<string, unknown> | undefined;
    if (usage) {
      const inputTokens = (usage.input_tokens as number) || 0;
      const outputTokens = (usage.output_tokens as number) || 0;
      span.tokens = {
        input: inputTokens,
        output: outputTokens,
        total: inputTokens + outputTokens,
      };

      const cost = calculateCost(model, inputTokens, outputTokens);
      if (cost !== null) {
        span.costUsd = cost;
      }
    }

    // Extract output content
    const content = obj.content as Array<Record<string, unknown>> | undefined;
    if (content && content.length > 0) {
      const textBlocks = content
        .filter((b) => b.type === "text")
        .map((b) => b.text as string);
      span.output = {
        content: textBlocks.join("\n"),
        stop_reason: obj.stop_reason,
      };
    }

    span.attributes["response_id"] = obj.id;
    span.attributes["model"] = obj.model;
    span.status = "ok";
  }

  span.endTime = Date.now() / 1000;
}

export function instrument(): void {
  try {
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    const anthropic = require("anthropic");
    if (anthropic?.Anthropic) {
      const Anthropic = anthropic.Anthropic;
      const proto = Anthropic.prototype;

      if (proto?.messages?.create) {
        originalMessagesCreate = proto.messages.create;
        proto.messages.create = function (
          this: unknown,
          ...args: unknown[]
        ) {
          const options = args[0] as Record<string, unknown> | undefined;
          const model = (options?.model as string) || "unknown";
          const messages = (options?.messages as unknown[]) || [];

          const span = buildSpan(model, messages);

          try {
            const result = originalMessagesCreate!.apply(this, args);
            // Handle both sync and async
            if (result && typeof (result as Promise<unknown>).then === "function") {
              return (result as Promise<unknown>).then(
                (resolved: unknown) => {
                  finishSpan(span, resolved, model);
                  return resolved;
                },
                (error: Error) => {
                  span.status = "error";
                  span.attributes["error.message"] = error.message;
                  span.endTime = Date.now() / 1000;
                  throw error;
                }
              );
            }
            finishSpan(span, result, model);
            return result;
          } catch (error) {
            span.status = "error";
            span.attributes["error.message"] =
              error instanceof Error ? error.message : String(error);
            span.endTime = Date.now() / 1000;
            throw error;
          }
        };
      }

      // Patch async client if available
      if (anthropic.AsyncAnthropic) {
        const asyncProto = anthropic.AsyncAnthropic.prototype;
        if (asyncProto?.messages?.create) {
          originalAsyncMessagesCreate = asyncProto.messages.create;
          asyncProto.messages.create = function (
            this: object,
            ...args: unknown[]
          ) {
            const options = args[0] as Record<string, unknown> | undefined;
            const model = (options?.model as string) || "unknown";
            const messages = (options?.messages as unknown[]) || [];

            const span = buildSpan(model, messages);
            const ctx = this;

            return (originalAsyncMessagesCreate!.apply(ctx, args) as Promise<unknown>).then(
              (result: unknown) => {
                finishSpan(span, result, model);
                return result;
              },
              (error: Error) => {
                span.status = "error";
                span.attributes["error.message"] = error.message;
                span.endTime = Date.now() / 1000;
                throw error;
              }
            );
          };
        }
      }
    }
  } catch {
    // Anthropic not available, skip instrumentation
  }
}

export function uninstrument(): void {
  try {
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    const anthropic = require("anthropic");
    if (anthropic?.Anthropic && originalMessagesCreate) {
      anthropic.Anthropic.prototype.messages.create = originalMessagesCreate;
    }
    if (anthropic?.AsyncAnthropic && originalAsyncMessagesCreate) {
      anthropic.AsyncAnthropic.prototype.messages.create = originalAsyncMessagesCreate;
    }
  } catch {
    // Ignore
  }
  originalMessagesCreate = null;
  originalAsyncMessagesCreate = null;
}
